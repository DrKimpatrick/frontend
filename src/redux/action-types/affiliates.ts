export enum AffiliateTypes {
  AddMoreInfo = 'Affiliate/AddMoreInfo',
  Errors = 'Affiliate/Errors',
  Loading = 'Affiliate/Loading'
}

// export interface CompanyOrSchool {
//   _id?: string;
//   name: string;
//   address: string;
//   website: string;
//   accountManagerName: string;
//   email: string;
//   phone: string;
//   userId?: string;
//   createdAt?: string;
//   updatedAt?: string;
// }

interface AddMoreInfo {
  type: typeof AffiliateTypes.AddMoreInfo;
  payload: {
    data: any;
  };
}

interface Errors {
  type: typeof AffiliateTypes.Errors;
  payload: {
    errors: any;
  };
}

interface Loading {
  type: typeof AffiliateTypes.Loading;
}

export type CompanyActionTypes = AddMoreInfo | Errors | Loading;
