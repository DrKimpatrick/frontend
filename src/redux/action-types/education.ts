export enum EducationTypes {
  AddEducation = 'Education/AddEducation',
  DeleteEducation = 'Education/DeleteEducation',
  UpdateEducation = 'Education/UpdateEducation',
  ListEducation = 'Education/ListEducation',
  ListSpecificEducation = 'Education/ListSpecificEducation',
  Errors = 'Education/Errors',
  Loading = 'Education/Loading',
  ChangeEducationStatus = 'Education/ChangeEducationStatus'
}

export enum VerificationStatus {
  Unverified = 'unverified',
  InProgress = 'inProgress',
  Verified = 'verified'
}
export interface Education {
  _id: string;
  schoolName: string;
  level: string;
  degreeOrCertification: string;
  specializations: string[];
  startDate: string;
  endDate: string;
  accomplishments: string[];
  isCurrentEducation: boolean;
  verificationStatus?: VerificationStatus;
  createdAt?: string;
  updatedAt?: string;
}

interface AddEducation {
  type: typeof EducationTypes.AddEducation;
  payload: {
    data: Education;
  };
}

interface DeleteEducation {
  type: typeof EducationTypes.DeleteEducation;
  payload: {
    id: string;
  };
}

interface UpdateEducation {
  type: typeof EducationTypes.UpdateEducation;
  payload: {
    data: Education;
  };
}

interface ListEducation {
  type: typeof EducationTypes.ListEducation;
  payload: {
    data: Education[];
  };
}

interface ListSpecificEducation {
  type: typeof EducationTypes.ListSpecificEducation;
  payload: {
    data: Education;
  };
}

interface Errors {
  type: typeof EducationTypes.Errors;
  payload: {
    errors: any;
  };
}

interface Loading {
  type: typeof EducationTypes.Loading;
  payload: {
    loading?: boolean;
  };
}

interface ChangeEducationStatus {
  type: typeof EducationTypes.ChangeEducationStatus;
  payload: {
    data: Education;
  };
}

export type EducationActionTypes =
  | AddEducation
  | DeleteEducation
  | UpdateEducation
  | ListEducation
  | ListSpecificEducation
  | Errors
  | Loading
  | ChangeEducationStatus;
