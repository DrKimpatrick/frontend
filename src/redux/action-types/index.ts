// This is where we will add action types