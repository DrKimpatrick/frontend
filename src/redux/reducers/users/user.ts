import { SkillTypes, SkillActionTypes } from 'redux/action-types/skill';
import {
  RegisterDispatchTypes,
  REGISTER_LOADING,
  REGISTER_FAIL,
  REGISTER_SUCCESS
} from 'redux/action-types/getStarted';

import {
  LoginDispatchTypes,
  LOGIN_LOADING,
  LOGIN_FAIL,
  LOGIN_SUCCESS
} from 'redux/action-types/login';

import {
  FORGOT_FAIL,
  FORGOT_LOADING,
  FORGOT_SUCCESS,
  ForgotDispatchTypes
} from 'redux/action-types/forgotPass';

import {
  UserTypeActions,
  UserTypes,
  User,
  UserSkill
} from 'redux/action-types/user';

import {
  RESET_FAIL,
  RESET_LOADING,
  RESET_SUCCESS,
  ResetDispatchTypes
} from 'redux/action-types/resetPass';

import {
  VERIFY_SUCCESS,
  VERIFY_LOADING,
  VERIFY_FAIL,
  VerifyDispatchTypes
} from 'redux/action-types/verifyAcc';

interface InitialState {
  currentUser: {
    isLoggedIn: boolean;
    data: object | null | any;
    loading: boolean;
    error?: any;
    message?: string;
    created?: boolean;
  };
  user?: User;
  loading?: boolean;
  errors?: any;
  specificUser?: User;
  userSkill?: UserSkill[];
  usersByRole?: {
    totalDocs?: number;
    users?: User[];
    currentPage?: number;
  };
  // skills?: AddedSkill[];
  activePath: string | null;
}

const initialState: InitialState = {
  currentUser: {
    isLoggedIn: !!localStorage.getItem('token'),
    loading: false,
    created: false,
    data: {
      firstName: 'John',
      lastName: 'Smith',
      profilePicture:
        'https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png'
    }
  },
  activePath: null
};

export const userReducer = (
  state = initialState,
  action:
    | RegisterDispatchTypes
    | LoginDispatchTypes
    | ForgotDispatchTypes
    | ResetDispatchTypes
    | VerifyDispatchTypes
    | UserTypeActions
    | SkillActionTypes
) => {
  switch (action.type) {
    case REGISTER_FAIL:
      return {
        ...state,
        errorSignup: action.payload,
        created: true,
        loading: false
      };

    case REGISTER_SUCCESS:
      return {
        ...state,
        loading: false,
        currentUser: {
          data: action.payload,
          isLoggedIn: false,
          created: true,
          error: null
        }
      };
    case LOGIN_LOADING:
      return {
        ...state,
        loading: true
      };
    case FORGOT_LOADING:
      return {
        ...state,
        loading: true
      };
    case RESET_LOADING:
      return {
        ...state,
        loading: true
      };
    case VERIFY_LOADING:
      return {
        ...state,
        loading: true
      };
    case REGISTER_LOADING:
      return {
        ...state,
        loading: true
      };
    case LOGIN_FAIL:
      return {
        ...state,
        errorLogin: action.payload,
        loading: false
      };

    case LOGIN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        loading: false
      };

    case FORGOT_FAIL:
      return {
        ...state,
        errorForgot: action.payload,
        loading: false
      };

    case FORGOT_SUCCESS:
      return {
        ...state,
        errorForgot: null,
        loading: false
      };

    case RESET_FAIL:
      return {
        ...state,
        errorReset: action.payload,
        loading: false
      };

    case RESET_SUCCESS:
      return {
        ...state,
        errorReset: null,
        resetSuccess: true,
        loading: false
      };

    case VERIFY_FAIL:
      return {
        ...state,
        errorVerify: action.payload,
        loading: false
      };

    case VERIFY_SUCCESS:
      return {
        ...state,
        message: action.payload,
        errorVerify: null,
        loading: false
      };

    case UserTypes.CurrentUser:
      return { ...state, user: action.payload.data, loading: false };

    case UserTypes.Errors:
      return { ...state, errors: action.payload.errors, loading: false };

    case UserTypes.Loading:
      return { ...state, loading: action.payload.loading, errors: null };

    case UserTypes.ListSpecificUser:
      return {
        ...state,
        specificUser: action.payload.data,
        loading: false,
        errors: undefined
      };

    case UserTypes.ListUserSkill:
      return { ...state, userSkill: action.payload.data, loading: false };

    case SkillTypes.GetSkillsFail:
      return { ...state, errors: action.payload, loading: false };

    case SkillTypes.GetSkillsLoading:
      return { ...state, loading: true };

    case SkillTypes.GetSkillsSuccess:
      return { ...state, skills: action.payload, loading: false, errors: null };

    case SkillTypes.SkillFail:
      return { ...state, errors: action.payload, loading: false };

    case SkillTypes.SkillSuccess:
      return {
        ...state,
        addedSkill: action.payload,
        loading: false,
        errors: null
      };

    case SkillTypes.SkillLoading:
      return { ...state, loading: true };

    case UserTypes.ListUserByRole:
      return {
        ...state,
        usersByRole: {
          users: action.payload.data.data,
          totalDocs: action.payload.data.totalDocs,
          currentPage: action.payload.data.currentPage
        },
        loading: false
      };

    case UserTypes.ActivePath:
      return { ...state, activePath: action.payload.data };

    case UserTypes.UpdateUser:
      return {
        ...state,
        user: action.payload.data,
        loading: false,
        errors: null
      };

    default:
      return state;
  }
};

export default userReducer;
