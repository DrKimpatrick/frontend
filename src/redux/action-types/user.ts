import { Employment } from './employment';
import { Education } from './education';
import { Skill } from './skill';

export enum UserTypes {
  CurrentUser = 'User/CurrentUser',
  Errors = 'User/Errors',
  ListSpecificUser = 'User/ListSpecificUser',
  Loading = 'User/Loading',
  ListUserSkill = 'User/ListUserSkill',
  ListUserByRole = 'User/ListUserByRole',
  ListTalentUser = 'User/ListTalentUser',
  ListHumanResourceUser = 'User/ListHumanResourceUser',
  ListAffiliateUser = 'User/ListAffiliateUser',
  ListHrAdminUser = 'User/ListHrAdminUser',
  ListTrainingUser = 'User/ListTrainingUser',
  ListRecruiterUser = 'User/ListRecruiterUser',
  ListCompanyUser = 'User/ListCompany',
  ActivePath = 'User/ActivePath',
  UpdateUser = 'User/UpdateUser',
  UserEducationLoading = 'User/UserEducationLoading',
  UserEmploymentLoading = 'User/UserEmploymentLoading',
  UserSkillLoading = 'User/UserSkillLoading',
  SetProfileProcess = 'User/SetProfileProcess'
}

export enum VerificationStatus {
  Verified = 'verified',
  Unverified = 'unverified',
  InProgress = 'inProgress'
}

export enum UserRole {
  Talent = 'talent',
  EducationUser = 'education',
  SuperAdmin = 'super_admin',
  RecruitmentAdmin = 'recruitment_admin',
  HrAdmin = 'hr_admin',
  CompanyAdmin = 'company_admin',
  TrainingAdmin = 'training_admin',
  TrainingAffiliate = 'training_affiliate'
}

export enum TalentProcess {
  CurrentRole = 'currentRole',
  SkillRanking = 'skillRanking',
  RecentEmployer = 'recentEmployer',
  AddEducation = 'addEducation',
  Completed = 'completed',
  SingleEducation = 'singleEducation',
  ListEducation = 'listEducation',
  SingleEmployment = 'singleEmployment',
  ListEmployment = 'listEmployment'
}

export enum AdminsProcess {
  AddCompany = 'AddCompany',
  AddSchool = 'AddSchool',
  AddPlan = 'AddPlan',
  Payment = 'Payment',
  Completed = 'Completed'
}

export interface User {
  _id: string;
  username: string;
  email: string;
  roles: string[];
  updatedAt: string;
  dateRegistered: string;
  paymentStatus: string;
  featureChoice: string;
  verified: boolean;
  employmentHistory?: Employment[];
  educationHistory?: Education[];
  profileProcess?: string;
  profilePicture?: string;
  bio?: string;
  stripeSubscriptionId?: string;
}

export interface UserSkill {
  _id: string;
  verificationStatus: string;
  level: string;
  skill: Skill;
  createdAt?: string;
  updatedAt?: string;
}

interface CurrentUser {
  type: typeof UserTypes.CurrentUser;
  payload: {
    data: User;
  };
}

interface Errors {
  type: typeof UserTypes.Errors;
  payload: {
    errors: any;
  };
}

interface Loading {
  type: typeof UserTypes.Loading;
  payload: {
    loading: boolean;
  };
}

interface ListSpecificUser {
  type: typeof UserTypes.ListSpecificUser;
  payload: {
    data: User;
  };
}

interface ListUserSkill {
  type: typeof UserTypes.ListUserSkill;
  payload: {
    data: UserSkill[];
  };
}

interface ListUserByRole {
  type: typeof UserTypes.ListUserByRole;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface ActivePath {
  type: typeof UserTypes.ActivePath;
  payload: {
    data: string;
  };
}

interface UpdateUser {
  type: typeof UserTypes.UpdateUser;
  payload: {
    data: User;
  };
}

interface ListTalentUser {
  type: typeof UserTypes.ListTalentUser;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface ListHrAdminUser {
  type: typeof UserTypes.ListHrAdminUser;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface ListRecruiterUser {
  type: typeof UserTypes.ListRecruiterUser;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface ListCompanyUser {
  type: typeof UserTypes.ListCompanyUser;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface ListAffiliateUser {
  type: typeof UserTypes.ListAffiliateUser;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface ListTrainingAffiliateUser {
  type: typeof UserTypes.ListTrainingUser;
  payload: {
    data: {
      totalDocs: number;
      data: User[];
      currentPage: number;
    };
  };
}

interface UserEducationLoading {
  type: typeof UserTypes.UserEducationLoading;
  payload: {
    loading: boolean;
  };
}

interface UserEmploymentLoading {
  type: typeof UserTypes.UserEmploymentLoading;
  payload: {
    loading: boolean;
  };
}

interface UserSkillLoading {
  type: typeof UserTypes.UserSkillLoading;
  payload: {
    loading: boolean;
  };
}

interface SetProfileProcess {
  type: typeof UserTypes.SetProfileProcess;
  payload: {
    data: string;
  };
}
export type UserTypeActions =
  | CurrentUser
  | Errors
  | ListSpecificUser
  | Loading
  | ListUserSkill
  | ListUserByRole
  | ActivePath
  | UpdateUser
  | ListTalentUser
  | ListAffiliateUser
  | ListTrainingAffiliateUser
  | ListRecruiterUser
  | ListCompanyUser
  | ListHrAdminUser
  | UserEducationLoading
  | UserEmploymentLoading
  | UserSkillLoading
  | SetProfileProcess;
