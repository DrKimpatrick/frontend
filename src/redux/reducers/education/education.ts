import { map } from 'lodash';
import { EducationTypes } from '../../actions/education';
import { MessageActionType } from '../../action-types/message';
import { EducationActionTypes, Education } from '../../action-types/education';
import { MessageTypes } from '../../actions/message';

interface InitialState {
  readonly message?: string;
  readonly errors?: any;
  readonly educations?: Education[];
  readonly education?: Education;
  readonly loading?: boolean;
}

const initialState: InitialState = {};

export const educationReducer = (
  state = initialState,
  action: EducationActionTypes | MessageActionType
) => {
  switch (action.type) {
    case EducationTypes.AddEducation:
      return {
        ...state,
        educations: state.educations && [
          ...state.educations,
          action.payload.data
        ],
        loading: false,
        education: action.payload.data
      };

    case EducationTypes.DeleteEducation:
      return {
        ...state,
        loading: false,
        educations:
          state.educations &&
          state.educations.filter(item => item._id !== action.payload.id)
      };

    case EducationTypes.ListEducation:
      return { ...state, loading: false, educations: action.payload.data };

    case EducationTypes.ListSpecificEducation:
      return { ...state, loading: false, education: action.payload.data };

    case EducationTypes.UpdateEducation:
      const update = map(state.educations, item => {
        if (item._id === action.payload.data._id) {
          item = action.payload.data;
        }
        return item;
      });
      return {
        ...state,
        loading: false,
        educations: state.educations && update
      };

    case EducationTypes.Errors:
      return { ...state, loading: false, errors: action.payload.errors };

    case EducationTypes.Loading:
      return {
        ...state,
        loading: action.payload.loading ? action.payload.loading : true
      };

    case MessageTypes.Error:
    case MessageTypes.Message:
      return { ...state, loading: false };

    default:
      return state;
  }
};
