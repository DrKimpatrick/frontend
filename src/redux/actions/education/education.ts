import { Dispatch } from 'redux';
import { InitialEducationValue as InitialValue } from 'components/Talent/Dashboard/Education';
import { EducationTypes } from 'redux/action-types/education';
import { UserTypes } from 'redux/action-types/user';
import { setProfileProcess } from 'redux/actions/user';
import ApiAction from '../../../helpers/apiAction';
import { setMessage } from '../message';

export const addEducation = (
  data: InitialValue,
  userInfo?: { userId: string; profileProcess: string }
) => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: '/education',
      method: 'POST',
      data,
      onStart: () => (dispatch: Dispatch) =>
        dispatch({ type: EducationTypes.Loading, payload: { loading: true } }),
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.Errors,
          payload: {
            errors: error.response.data
          }
        });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        if (userInfo && userInfo.profileProcess && userInfo.userId) {
          setProfileProcess({
            userId: userInfo.userId,
            profileProcess: userInfo.profileProcess
          })(dispatch);
        }
        dispatch({
          type: EducationTypes.AddEducation,
          payload: { data: res.data }
        });
        setMessage('education saved successfully')(dispatch);
      }
    })
  );
};

export const listEducation = () => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: '/education',
      method: 'GET',
      onStart: () => (dispatch: Dispatch) => {
        dispatch({ type: EducationTypes.Loading, payload: { loading: true } });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.Errors,
          payload: {
            errors: error.response.data
          }
        });
        dispatch({ type: EducationTypes.Loading, payload: { loading: false } });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({ type: EducationTypes.Loading, payload: { loading: false } });
        if (res.data) {
          dispatch({
            type: EducationTypes.ListEducation,
            payload: {
              data: res.data
            }
          });
        }
      }
    })
  );
};

export const listSpecificEducation = (userId: string, educationId: string) => (
  dispatchAction: Dispatch
) => {
  return dispatchAction(
    ApiAction({
      url: `/education/${userId}/${educationId}`,
      method: 'GET',
      onStart: () => (dispatch: Dispatch) =>
        dispatch({ type: EducationTypes.Loading, payload: { loading: true } }),
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.Errors,
          payload: {
            errors: error.response.data
          }
        });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.ListSpecificEducation,
          payload: { data: res.data }
        });
      }
    })
  );
};

export const updateEducation = (
  data: InitialValue,
  userId: string,
  educationId: string
) => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: `/education/${userId}/${educationId}`,
      method: 'PATCH',
      data,
      onStart: () => (dispatch: Dispatch) =>
        dispatch({ type: EducationTypes.Loading, payload: { loading: true } }),
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.Errors,
          payload: {
            errors: error.response.data
          }
        });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.UpdateEducation,
          payload: { data: res.data }
        });
        setMessage('education updated successfully')(dispatch);
      }
    })
  );
};

export const deleteEducation = (educationId: string) => (
  dispatchAction: Dispatch
) => {
  return dispatchAction(
    ApiAction({
      url: `/education/${educationId}`,
      method: 'Delete',
      onStart: () => (dispatch: Dispatch) =>
        dispatch({ type: EducationTypes.Loading, payload: { loading: true } }),
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.Errors,
          payload: {
            errors: error.response.data
          }
        });
      },
      onSuccess: () => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.DeleteEducation,
          payload: { id: educationId }
        });
        setMessage('education deleted successfully')(dispatch);
      }
    })
  );
};

export const changeEducationStatus = (values: {
  status: string;
  id: string;
}) => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: `/education/status/${values.id}`,
      method: 'Put',
      data: { verificationStatus: values.status },
      onStart: () => dispatch => {
        dispatch({
          type: UserTypes.UserEducationLoading,
          payload: { loading: true }
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.Errors,
          payload: {
            errors: error.response.data
          }
        });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: EducationTypes.ChangeEducationStatus,
          payload: { data: res.data }
        });
        setMessage(res.message)(dispatch);
      }
    })
  );
};
