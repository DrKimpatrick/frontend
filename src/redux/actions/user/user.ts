import { Dispatch } from 'redux';
import { UserTypes } from 'redux/action-types/user';
import { AddSkill, SkillTypes } from 'redux/action-types/skill';
import ApiAction from '../../../helpers/apiAction';

export const listCurrentUser = () => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: '/users/me',
      method: 'GET',
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.CurrentUser,
          payload: { data: res.profile }
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.Errors,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};

export const listSpecificUser = (userId: string) => (
  dispatchAction: Dispatch
) => {
  return dispatchAction(
    ApiAction({
      url: `/users/${userId}`,
      method: 'GET',
      onStart: () => (dispatch: Dispatch) => {
        dispatch({ type: UserTypes.Loading, payload: { loading: true } });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.ListSpecificUser,
          payload: { data: res.profile }
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.Errors,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};

export const listUserSkill = (userId: string) => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: `/users/${userId}/skills`,
      method: 'GET',
      onStart: () => (dispatch: Dispatch) => {
        dispatch({ type: UserTypes.Loading, payload: { loading: true } });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.ListUserSkill,
          payload: { data: res.data }
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.Errors,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};

export const listAllSkill = () => (dispatchAction: Dispatch) => {
  return dispatchAction(
    ApiAction({
      url: `/skills`,
      method: 'GET',
      onStart: () => (dispatch: Dispatch) => {
        dispatch({
          type: SkillTypes.GetSkillsLoading,
          payload: { loading: true }
        });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: SkillTypes.GetSkillsSuccess,
          payload: res.data
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: SkillTypes.GetSkillsFail,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};

export const listUserByRole = (data: {
  page: number;
  limit: number;
  role: string;
}) => (dispatchAction: Dispatch) => {
  const { page, limit, role } = data;
  return dispatchAction(
    ApiAction({
      url: `/users/?roles=${role}&limit=${limit}&page=${page}`,
      method: 'GET',
      onStart: () => (dispatch: Dispatch) => {
        dispatch({ type: UserTypes.Loading, payload: { loading: true } });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.ListUserByRole,
          payload: { data: res }
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.Errors,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};

export const addedSkillsUser = (dataValues: AddSkill[]) => (
  dispatchAction: Dispatch
) => {
  console.log('data values', dataValues);
  return dispatchAction(
    ApiAction({
      url: '/skills/me',
      method: 'POST',
      data: dataValues,
      onStart: () => (dispatch: Dispatch) => {
        dispatch({
          type: SkillTypes.SkillLoading,
          payload: { loading: true }
        });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        dispatch({
          type: SkillTypes.SkillSuccess,
          payload: { data: res }
        });
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: SkillTypes.SkillFail,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};

export const setActivePath = (pathname: string) => (dispatch: Dispatch) => {
  dispatch({ type: UserTypes.ActivePath, payload: { data: pathname } });
};

export const updateUser = (data: any, userId: string) => (
  dispatchAction: Dispatch
) => {
  return dispatchAction(
    ApiAction({
      url: `/users/${userId}`,
      method: 'PATCH',
      data,
      onStart: () => (dispatch: Dispatch) => {
        dispatch({ type: UserTypes.Loading, payload: { loading: true } });
      },
      onSuccess: res => (dispatch: Dispatch) => {
        delete res.profile.__v;
        setTimeout(() => {
          dispatch({
            type: UserTypes.UpdateUser,
            payload: { data: res.profile }
          });
        }, [1000]);
      },
      onFailure: error => (dispatch: Dispatch) => {
        dispatch({
          type: UserTypes.Errors,
          payload: { errors: error.response.data }
        });
      }
    })
  );
};
