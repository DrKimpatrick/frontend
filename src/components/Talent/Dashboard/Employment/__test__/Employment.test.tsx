import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import Component from '../Employment';
import { cleanup, render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { VerificationStatus } from 'redux/action-types/education';
import { UserRole } from 'redux/action-types/user';

let store: Store;

const middleware = [thunk];

const mockStore = configureMockStore(middleware);

const employmentState = {
  _id: '_id',
  companyName: 'companyName',
  supervisor: {
      name: 'Employee',
      detail: {name: 'grma', email: 'email@gmail.com', phoneNumber: '0786601003'}
  },
  skillsUsed: [],
  responsibilities: [],
  accomplishments: [],
  startDate: '2020-10-10',
  endDate: '2020-12-12',
  isCurrentPosition: false,
  title: 'Software Engineer',
  favoriteProject: '',
  verificationStatus: VerificationStatus.Unverified
};

const initialState = {
  employments: {
    employments: [employmentState],
    loading: false,
    employment: employmentState,
  },
  users: {
    loading: false,
    user: {
      id: 'id',
      username: 'username',
      email: 'email@gmail.com',
      employmentHistory: [employmentState],
      educationHistory: [],
      isVerified: true,
      roles: [UserRole.Talent],
      signupMode: 'Local'
    },
    userSkill: []
  },
  messages: {
    message: 'message',
    error: 'error'
  }
};

describe('Education', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });

  afterEach(cleanup);

  it('Should render without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(
      <Provider store={store}>
        <Router>
          <Component />
        </Router>
      </Provider>,
      div
    );
  });

  it('should create snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Component />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should add employment', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Component />
        </Router>
      </Provider>
    );

    const button: Element | any = container.querySelector('.addEmployment');

    fireEvent.click(button);

    expect(button).toBeTruthy();
  });

  it('should set action to true when mouseover event triggered', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Component />
        </Router>
      </Provider>
    );

    const li: Element | any = container.querySelector('.action');

    fireEvent.mouseOver(li);

    expect(li).toBeTruthy();
  });

  it('should set action to false when mouseleave event trigger', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Component />
        </Router>
      </Provider>
    );
    

    const li: Element | any = container.querySelector('.action');

    fireEvent.mouseLeave(li);

    expect(li).toBeTruthy();
  });

  it('should edit employment', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Component />
        </Router>
      </Provider>
    );

    const li: Element | any = container.querySelector('.action');

    fireEvent.mouseOver(li);

    const button: Element | any = container.querySelector('.editEmployment');

    fireEvent.click(button);

    expect(button).toBeTruthy();
  });

  it('should delete employment', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Component />
        </Router>
      </Provider>
    );

    const li: Element | any = container.querySelector('.action');

    fireEvent.mouseOver(li);

    const button: Element | any = container.querySelector('.delete');

    fireEvent.click(button);

    expect(button).toBeTruthy();
  });

  describe('Loading', () => {
    beforeEach(() => {
      store = mockStore({
        ...initialState,
        employments: { ...initialState.employments, loading: true }
      });
    });
    it('should return loading component', () => {
      const { container } = render(
        <Provider store={store}>
          <Router>
            <Component />
          </Router>
        </Provider>
      );

      const div: Element | any = container.querySelector('.sideLoading');

      expect(div).toBeDefined();
    });
  });

  describe('NotFound', () => {
    beforeEach(() => {
      store = mockStore({
        ...initialState,
        employments: { ...initialState.employments, employments: [] }
      });
    });
    it('should return not found message', () => {
      const { container } = render(
        <Provider store={store}>
          <Router>
            <Component />
          </Router>
        </Provider>
      );

      const h5: Element | any = container.querySelector('.notFound h5');

      expect(h5).toHaveTextContent('You currently do not have employment history');
    });
  });
});
