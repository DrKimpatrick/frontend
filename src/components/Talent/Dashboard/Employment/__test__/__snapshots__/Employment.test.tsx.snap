// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Education should create snapshot 1`] = `
<div
  className="w-full"
>
  <p
    className="text-right text-xs py-1 pb-2"
  />
  <ul
    className="p-0 bg-card-preview"
  >
    <li
      className="py-2 py-2 rounded-sm action"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="flex flex-row text-sm justify-between"
      >
        <h4
          className="medium-fonts px-4"
        >
          companyName
        </h4>
        <div
          className="flex flex-row"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root not-verified-color mr-4"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-7 9h-2V5h2v6zm0 4h-2v-2h2v2z"
            />
          </svg>
        </div>
      </div>
    </li>
  </ul>
  <div
    className="addButton"
  >
    <button
      className="mt-4 bg-gray-800 w-full text-white hover:bg-gray-900 outline-none font-semibold h-12 py-1 px-1 rounded-sm shadow addEmployment"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
       
      <span>
        Add new Employment
      </span>
    </button>
  </div>
</div>
`;
